client_side/.infrastructure/infrastructure.sh

```diff
@@ -0,0 +1,201 @@
+get_task_code() {
+  if ((task_set == 1)); then
+    local task_no=$((curr_task > TASKS_SIZE / 2 ? \
+      curr_task - TASKS_SIZE / 2 : \
+      curr_task))
+  else
+    local task_no=$((curr_task > TASKS_SIZE / 2 ? \
+      curr_task : \
+      curr_task + TASKS_SIZE / 2))
+  fi
+
+
+  echo "$(char_from ${task_no})"
+}
+
+# Enables Bash preexec functions and prints out the first treatment and task.
+# Also sets $SECONDS and $time_elapsed to 0.
+start_experiment() {
+  # Enable task logging
+  preexec_functions+=(preexec_func)
+  precmd_functions+=(precmd_func)
+
+  print_treatment
+  print_task
+
+  SECONDS=0
+  time_elapsed=0
+}
+
+# "Uninstalls" Bash Preexec by removing its triggers.
+# Remove all variable files created by the infrastructure
+# Stops the experiment completely by returning from the sourced scripts.
+end_experiment() {
+  # This effectively uninstalls Bash Pre-exec
+  # Makes it so any commands typed after the experiment has ended will not be
+  # passed through preexec and precmd.
+  PROMPT_COMMAND=${PROMPT_COMMAND_OG}
+  trap - DEBUG
+
+  # Remove all variable files
+  rm -r "${INFRA_DIR}/.*"
+  cd "${EXP_DIR}"
+
+  echo "Congratulations! You have completed the interactive portion of the" \
+    "experiment!"
+  echo "Please take some time to fill out the survey at ${SURVEY_URL} ."
+
+  return 0
+}
+
+# Creates the file system directory that the user will be working on.
+# If the directory already exists, removes it and creates it again.
+# Saves the user's current working directory and returnst to it once the
+# directory is set up.
+make_fs() {
+  pushd "${INFRA_DIR}" &> /dev/null
+
+  rm -rf "${FS_DIR}"
+  mkdir "${FS_DIR}" > /dev/null
+  tar -xzf "${INFRA_DIR}/fs.tgz" -C "${FS_DIR}"
+
+  # Moves the user to this directory
+  popd &> /dev/null
+}
+
+# Prints the list of resources that the user is allowed to use based on the
+# current treatment
+print_treatment() {
+  echo -n "For this half of the experiment you can use any online resources," \
+    "man pages, "
+  if [[ "$(cat "${INFRA_DIR}/.treatment")" == "T" ]]; then
+    echo "and Tellina <URL> to help you solve the tasks."
+  else
+    echo "but you can't use Tellina to help you solve the tasks."
+  fi
+}
+
+# Prints the current task number and its description.
+print_task() {
+  local task_code=$(cat "${INFRA_DIR}/.task_code")
+
+  echo "-----------------------------------------------------------------------"
+  echo "Task: ${curr_task}/${TASKS_SIZE}"
+
+  jq -r '.description' "${TASKS_DIR}/task_${task_code}/task_${task_code}.json"
+}
+
+# See documentation for ./verify_task.py for more details on what it does
+# Captures the stdout of verify_task.py into $status
+# Captures the exit code of verify_task.py into $EXIT
+verify_task() {
+  # Verify the output of the previous command.
+  local task_code="$(cat "${INFRA_DIR}/.task_code")"
+  local user_command="$(cat "${INFRA_DIR}/.command")"
+
+  status=$("${INFRA_DIR}"/verify_task.py ${task_code} ${user_command})
```

Comment:
I am concerned about the lack of quoting of `${user_command}` on this line.
Quoting commands correctly to pass them to another program is a bit of a mess.  Should `verify_task.py` take a filename instead of (or as an alternate to) the command itself?

Reply:
The current implementation for `verify_task.py` is that every argument after the first (i.e. 2, 3, ...) is part of the user's command. With that said, I don't see a case where that would be a problem. 

If the user command is properly quoted and escaped, so should that passed to `verify_task.py`. If it is not, then there would be an error in the shell, and `verify_task.py` should fail as well.

Reply:
I know I've been bitten by this in the past.  I have attached [a small test case](https://github.com/anhnamtran/tellina_user_experiment/files/3148745/bash_test.zip).  It should print 3 lines, but instead it prints 4.  
Here is the result of executing it:

```
% bash_test.sh
echo
bbb
"ccc
ddd"
```

Are you sure that your code will not be susceptible to a problem like this?
You might want to store each argument on its own line in .command, unquoted.